### PR TITLE
Deprecate {{with}}

### PR DESCRIPTION
**Pull request by GavinJoyce**
closes https://github.com/emberjs/rfcs/issues/316

[Rendered](https://github.com/gavinjoyce/rfcs/blob/gj/deprecate-with/text/0000-template.md)

### CODE REVIEW

**Comment by rwjblue** [comment]
Definitely in favor of migrating `{{with` -> `{{if-let`, thanks for picking this up @GavinJoyce!

File: text/0000-template.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+We'll mentiton the new `{{if-let}}` helper in an Ember point release blog post.
+
+The deprecation message will link to good documentation which describe why `{{with}}` was deprecated and gives clear guidelines on how to migrate to using either `{{let}}` or `{{if-let}}`.
+
+## Drawbacks
+
+This adds a litte API and documentation churn.
+
+## Alternatives
```
Comment (patocallaghan) [inline text/0000-template.md]: There's probably a very good reason for not doing it but just throwing this out there. Rather than increase the template API surface area could we enhance `{{let}}` with the `{{else}}` functionality?

Comment (GavinJoyce) [reply text/0000-template.md]: @patocallaghan see the [motivation section](https://github.com/emberjs/rfcs/blob/master/text/0286-block-let-template-helper.md#motivation) from the `{{let}}` RFC:

> While the conditional semantics of with are coherent with the other built-in helpers like each and if, users often find this unexpected. The fact that only the first positional parameter of with controls whether the block is rendered might also add to the confusion.

`{{let}}` was purposefully designed to always render its block.

**Comment by tcjr** [comment]
Deprecating `{{with}}` is a good idea.  But `{{if-let}}` seems completely superfluous since you can get the desired behavior with `{{let}}` and `{{if}}`.  

```
{{#let "Alex" as |value|}}
  {{#if value}}
    {{value}} is truthy
  {{else}}
    The first positional param was falsy
  {{/if}}
{{/let}}
```

If the legacy `{{with}}` didn’t exist, would we even consider adding `{{if-let}}` to core?  

What about `{{unless-let}}`?

**Comment by hemlok** [comment]
> But `{{if-let}}` seems completely superfluous since you can get the desired behavior with `{{let}}` and `{{if}}`.

Another benefit of composing {{#let}} and nested {{#if}} is that you could handle multiple params and react to different positional params being false-y in different ways in the UI 🤔

**Comment by GavinJoyce** [comment]
@tcjr Thanks, I'll add your suggestion to the list of alternatives. I crafted this RFC based on the comments in https://github.com/emberjs/rfcs/issues/316#issuecomment-449489686, it seemed to indicate the the core team where in favor of adding `{{if-let}}`.

I actually prefer your example of combining `{{let}}` and `{{if}}`, I think it reads better. Also, if we were to add `{{if-let}}`, I don't believe that it would be widely used. I can't think of a place in our own app where we would us it.

> This just came up again at a core team call, I think we are all 👍 to adding `if-let` (with today's `{{with}}` semantics) and deprecating `{{with}}` itself.

@rwjblue do you remember why there was support from the core team for adding `{{if-let}}`. Was a combination of `if` and `let` also considered?

**Comment by GavinJoyce** [comment]
Thinking about it more, I strongly agree that we should just deprecate `{{with}}` and not introduce `{{let-if}}`. I'll update the RFC to reflect this now. If some good reasons emerge to introduce `{{let-if}}` and there is consensus, I can add it back to the RFC.

**Comment by kellyselden** [comment]
@GavinJoyce Perhaps you could also alter the RFC to ensure public API exists so that and addon can implement `if-let` without hacks (might already be possible).

**Comment by GavinJoyce** [comment]
@kellyselden here's a simple `{{if-let}}` component implementation:

https://ember-twiddle.com/9af662e85c5f06b6e4cb7f58ab08c954?openFiles=templates.application.hbs%2Ctemplates.components.if-let.hbs

screen shot 2019-02-14 at 13 27 26

**Comment by GavinJoyce** [comment]
😜I got the naming wrong in the twiddle above, used `{{if-let}}` instead of `{{let-if}}`. I think that's another good reason not to introduce `{{let-if}}` in ember core.

**Comment by rwjblue** [comment]
I think that @tcjr / @hemlok's intuition here is good. Manually writing out the nested version of if-let is easy enough, is not much longer than the shorthand `{{if-let` version, and is _probably_ more intuitive for folks that don't have experience with languages that themselves have a built in `if-let` (e.g. rust or some lisps).

tldr; 👍 on just making this a straight up deprecation RFC...

File: text/0000-template.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  {{/if}}
+{{/let}}
+```
+
+## Detailed design
+
+We'll create an AST transform in [`packages/ember-template-compiler`](https://github.com/emberjs/ember.js/tree/master/packages/ember-template-compiler) which will emit a deprecation warning for all uses of `{{with}}`. The deprecation warning will be something like:
+
+```
+Using `{{with}}` is deprecated, please use a combination of `{{let}}` and `{{if}}` instead.
```
Comment (rwjblue) [inline text/0000-template.md]: I think that the message should only suggest `if` _if_ there is an `{{else`, what do you think?

File: text/0000-template.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+```
+
+This new deprecation warning should link to some documentation which clearly describes when you should use `{{let}}` and when you should use `{{let}}` and `{{if}}` in combination.
+
+## How we teach this
+
+We'll mentiton the deprecation in an Ember point release blog post.
+
+The deprecation message will link to good documentation which describe why `{{with}}` was deprecated and gives clear guidelines on how to migrate to using either `{{let}}` or `{{let}}` and `{{it}}` in combination.
```
Comment (rwjblue) [inline text/0000-template.md]: We should include the deprecation guide prose in this section directly. The deprecation guide entry will be linked to directly from the deprecation itself, and should show before/after examples for the various styles of `{{with` usage.

File: text/0000-template.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+We could leave `{{with}}` as is. I don't believe that this is a good option as the name `{{with}}` is confusing.
+
+We could deprecate `{{with}}` and introduce `{{if-let}}` in Ember core. This RFC originally made that exact proposal, I was strongly persuaded of the [lack of need for `{{if-let}}` by @tcjr](https://github.com/emberjs/rfcs/pull/445#issuecomment-463594185).
+
+We could deprecate `{{with}}` and introduce `{{if-let}}` in an addon instead of Ember core.
+
+## Unresolved questions
+
+This [comment](https://github.com/emberjs/rfcs/issues/316#issuecomment-449489686) seems to indicate some support for `{{if-let}}` from the core team. Perhaps there are some compelling reasons for add it that are not currently included in this RFC?
```
Comment (rwjblue) [inline text/0000-template.md]: I _personally_ like `if-let`, but I think that this is just as easy to understand and doesn't rely on the reader to grok extra concepts....

File: text/0000-template.md (hunk 2 above)
Comment (GavinJoyce) [reply text/0000-template.md]: Perhaps some people rely on the main `{{with}}` block not rendering when the first argument in falsy? If so, I believe that they would need to migrate to using `{{let}}` and `{{if}}`.

**Comment by dgeb** [comment]
Good discussion. I also favor deprecating `{{with}}` without introducing a new replacement simultaneously.

File: text/0000-template.md (hunk 2 above)
Comment (rwjblue) [reply text/0000-template.md]: Ya, I totally agree. My _guess_ (and anecdotally what I've found when teaching/pairing) is that most folks are **surprised** by the current behavior and actually do want "render this block all the time".

I think I'd propose the following "choose your own adventure" for the deprecation guide on migrating away from `{{with` usage:

* if you always want to render a given block unconditionally: use `{{let`
* if you want to render a given block conditionally: use `{{let` + `{{if`
* if you want to render a given block conditionally, otherwise rendering an alternate block: use `{{let` + `{{if` + `{{else`

**Comment by rwjblue** [comment]
I'd also like to suggest creating a linting rule (to aid folks who are not yet on newer versions with the deprecation but can use `{{let`, to avoid `{{with`) and a codemod (leveraging `codemod-cli` + `ember-template-recast` _should_ make this straight forward).

File: text/0000-template.md (hunk 2 above)
Comment (GavinJoyce) [reply text/0000-template.md]: 👍 sounds good, thanks

Comment (GavinJoyce) [reply text/0000-template.md]: @rwjblue I've added the deprecation guide content to this RFC

**Comment by GavinJoyce** [comment]
> I'd also like to suggest creating a linting rule

✅ 

> and a codemod (leveraging `codemod-cli` + `ember-template-recast` _should_ make this straight forward).

@rwjblue for the same reasons as why we need to provide three options in the deprecation guide, I'm not sure that it will be possible to fully automate the migration from `{{with}}` to `{{let}}`

**Comment by rwjblue** [comment]
> @rwjblue for the same reasons as why we need to provide three options in the deprecation guide, I'm not sure that it will be possible to fully automate the migration from `{{with}}` to `{{let}}`

Firstly, there _is_ a fully correct codemod that we could write. It would just look hideous in most cases when you _didn't_ want the conditional behaviors.

For example:

```hbs
{{#with (concat first ' ' last) as |name|}}
  <h2>Hello, {{name}}!</h2>
{{/with}}
```

Would have to be written as:

```hbs
{{#let (concat first ' ' last) as |name|}}
  {{#if name}}
    <h2>Hello, {{name}}!</h2>
  {{/if}}
{{/let}}
```

Even though we _mostly_ know that we don't need that conditional.

---

I think the codemod could actually have some pretty nice heuristics to be _pretty_ close to what you would do by hand:

* if the `{{#with` has an `{{else`, always emit the nested `{{#if`
* if the argument to `{{#with` is a subexpression, assume that it is non-falsey (and therefore do not emit the nested `{{#if` inside the block)
* otherwise use a "simple" `{{#let` without nested if

In addition, we could allow a flag to the codemod so that the person running it can opt in to "100% correctness, but super verbose" mode. The [Ember.K](https://emberjs.github.io/rfcs/0178-deprecate-ember-k.html) codemod did essentially the same thing by _requiring_ a flag to assume "chaining" or not ([codemod docs here](https://github.com/abuiles/ember-watson#remove-usages-of-emberk)).

**Comment by GavinJoyce** [comment]
> I think the codemod could actually have some pretty nice heuristics to be _pretty_ close to what you would do by hand

👍thanks, I've added a short note on adding a codemod to assist the migration

File: text/0000-template.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+For people on older versions of Ember that support `{{let}}`, we'll create an `ember-template-lint` rule that they can use to prevent the use of `{{with}}`.
+
+We'll also create a codemod which will assist people when migrating from `{{with}}` to `{{let}}`.
+
+## How we teach this
+
+We'll mentiton the deprecation in an Ember point release blog post.
+
+As mentioned above, the deprecation message will contain a link to clear guidelines on how to migrate to using `{{let}}`.
+
```
Comment (jenweber) [inline text/0000-template.md]:
```suggestion

There is nothing to remove from the Ember.js Guides since we already teach only the use of `{{let}}`.
```

**Comment by chancancode** [comment]
We discussed this today at the framework core team meeting and decided to move this RFC into Final Comment Period

File: text/0000-template.md (hunk 5 above)
Comment (GavinJoyce) [reply text/0000-template.md]: thanks

**Comment by chancancode** [comment]
We decided to accept this RFC at the framework core team meeting today. Thanks @GavinJoyce for working on this and everyone else for the discussion 🎉

**Comment by chancancode** [comment]
@GavinJoyce do you mind renaming the file to match the RFC number?

**Comment by GavinJoyce** [comment]
@chancancode updated, thanks
